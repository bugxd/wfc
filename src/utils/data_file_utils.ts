import { Tile, Tiles } from "../types";
import { DataFile, DataTile } from "../types/data_file_types";

export function toDataFile(props: {cellSize: number, tiles: Tiles, frequencies: number[]}): DataFile {
  const tileMap = new Map<string, Tile[]>();

  // find same tiles and add to tile map
  for(var i = 0; i < props.tiles.length; i++) {
    const tile = props.tiles[i];

    if(tileMap.has(tile.uid))
      tileMap.set(tile.uid, [...tileMap.get(tile.uid)!, tile]);
    else
      tileMap.set(tile.uid, [tile]);
  }

  const dataTiles : DataTile[] = [];

  // make one tile out of the same tiles
  tileMap.forEach((tiles: Tile[]) => {
    tiles.sort((a,b) => a.rotiationDegree - b.rotiationDegree);

    const zeroDegreeTile = tiles.shift();

    if(zeroDegreeTile) {
      // get frequency for tile
      let frequency = 0;
      const id = props.tiles.findIndex((t: Tile) => t.uid === zeroDegreeTile.uid);
      console.log(zeroDegreeTile.uid+ " " + id);
      if(id !== -1)
        frequency = props.frequencies[id];

      dataTiles.push({
        svg: zeroDegreeTile.svg,
        north: zeroDegreeTile.north,
        east: zeroDegreeTile.east,
        south: zeroDegreeTile.south,
        west: zeroDegreeTile.west,
        rotiationDegrees: [0, ...tiles.map(t => t.rotiationDegree)],
        frequency: frequency,
      });
    }
  });

  return {
    cellSize: props.cellSize,
    tiles: dataTiles,
  };
}

export function fromDataFile(content: string): {cellSize: number, tiles: Tiles, frequencies: number[]} {
  let cellSize: number = 0;
  const tiles:Tiles = [];
  let frequencies: number[] = [];

  const dataContent = JSON.parse(content) as DataFile;

  // test for completness
  if(!dataContent.cellSize) throw new Error("[fromDataFile]: missing cell size");
  if(!dataContent.tiles) throw new Error("[fromDataFile]: missing tiles");

  cellSize = dataContent.cellSize;
  for(var i = 0; i < dataContent.tiles.length; i++) {
    const tile: DataTile = dataContent.tiles[i];
    toTiles(tile).forEach(element => {
      tiles.push(element);
      frequencies.push(tile.frequency);
    });
  }

  return {cellSize: cellSize, tiles: tiles, frequencies: frequencies};
}

function toTiles(dataTile: DataTile): Tile[] {
  const uid = makeTileUid();
  return dataTile.rotiationDegrees.map(degree => ({
    svg: dataTile.svg,
    ...rotateDirections(degree, dataTile),
    rotiationDegree: degree,
    uid: uid,
  }));
}

function rotateDirections(degree:number, dataTile: DataTile): {north: string; east: string; south: string; west: string;} {
  switch(degree) {
    case 0: return {
      north: dataTile.north,
      east: dataTile.east,
      south: dataTile.south,
      west: dataTile.west,
    }
    case 90: return {
      north: dataTile.west.split("").reverse().join(""),
      east: dataTile.north,
      south: dataTile.east.split("").reverse().join(""),
      west: dataTile.south,
    }
    case 180: return {
      north: dataTile.south.split("").reverse().join(""),
      east: dataTile.west.split("").reverse().join(""),
      south: dataTile.north.split("").reverse().join(""),
      west: dataTile.east.split("").reverse().join(""),
    }
    case 270: return {
      north: dataTile.east,
      east: dataTile.south.split("").reverse().join(""),
      south: dataTile.west,
      west: dataTile.north.split("").reverse().join(""),
    }
    default:  return {
      north: dataTile.north,
      east: dataTile.east,
      south: dataTile.south,
      west: dataTile.west,
    }
  }
}

export function makeTileUid(): string {
    var length = 10;
    var result           = '';
    var characters       = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    var charactersLength = characters.length;
    for (var i = 0; i < length; i++) {
        result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
}
