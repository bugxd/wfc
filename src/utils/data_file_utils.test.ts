import { toDataFile, fromDataFile } from "./data_file_utils";

describe("data utils tests", () => {
  describe("toDataFile", () => {
    it("should combine same tiles", () => {
      const result = toDataFile({
        cellSize: 10,
        tiles: [{
          svg: "svg_a",
          north: "north_a",
          east: "east_a",
          south: "south_a",
          west: "west_a",
          rotiationDegree: 0,
          uid: "a",
        },{
          svg: "svg_a",
          north: "north_a",
          east: "east_a",
          south: "south_a",
          west: "west_a",
          rotiationDegree: 90,
          uid: "a",
        },{
          svg: "svg_a",
          north: "north_a",
          east: "east_a",
          south: "south_a",
          west: "west_a",
          rotiationDegree: 180,
          uid: "a",
        },{
          svg: "svg_a",
          north: "north_a",
          east: "east_a",
          south: "south_a",
          west: "west_a",
          rotiationDegree: 270,
          uid: "a",
        },{
          svg: "svg_b",
          north: "north_b",
          east: "east_b",
          south: "south_b",
          west: "west_b",
          rotiationDegree: 0,
          uid: "b",
        },{
          svg: "svg_b",
          north: "north_b",
          east: "east_b",
          south: "south_b",
          west: "west_b",
          rotiationDegree: 90,
          uid: "b",
        },{
          svg: "svg_c",
          north: "north_c",
          east: "east_c",
          south: "south_c",
          west: "west_c",
          rotiationDegree: 90,
          uid: "c",
        }],
        frequencies: [1,1,1,1,2,2,3]
      });

      expect(result).toMatchSnapshot();
    })
  });

  describe("fromDataFile", () => {
    beforeEach(() => {
      jest.spyOn(global.Math, 'random').mockReturnValue(0.234567891);
    });

    afterEach(() => {
        jest.spyOn(global.Math, 'random').mockRestore();
    });

    it("should throw error without cell size", () => {
      const content = JSON.stringify({
        tiles: [],
      })

      expect(() => fromDataFile(content)).toThrowError("[fromDataFile]: missing cell size");
    });

    it("should throw error without tiles", () => {
      const content = JSON.stringify({
        cellSize: 10,
      })

      expect(() => fromDataFile(content)).toThrowError("[fromDataFile]: missing tiles");
    });

    it("should get tiles from data", () => {
      const content = JSON.stringify({
        cellSize: 10,
        tiles: [{
          svg: "svg1",
          north: "north1",
          east: "east11",
          south: "south1",
          west: "west1",
          rotiationDegrees: [0, 90, 180, 270],
          frequency: 1,
        },{
          svg: "svg2",
          north: "north2",
          east: "east2",
          south: "south2",
          west: "west2",
          rotiationDegrees: [0, 90, 180, 270],
          frequency: 2,
        }],
      })

      const result = fromDataFile(content);

      expect(result.cellSize).toBe(10);
      expect(result.frequencies).toStrictEqual([1,1,1,1,2,2,2,2]);
      expect(result.tiles).toMatchSnapshot();
    });
  });
});
