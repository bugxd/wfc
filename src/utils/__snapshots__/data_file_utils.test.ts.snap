// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data utils tests fromDataFile should get tiles from data 1`] = `
Array [
  Object {
    "east": "east11",
    "north": "north1",
    "rotiationDegree": 0,
    "south": "south1",
    "svg": "svg1",
    "uid": "OOOOOOOOOO",
    "west": "west1",
  },
  Object {
    "east": "north1",
    "north": "west1",
    "rotiationDegree": 90,
    "south": "east11",
    "svg": "svg1",
    "uid": "OOOOOOOOOO",
    "west": "south1",
  },
  Object {
    "east": "west1",
    "north": "south1",
    "rotiationDegree": 180,
    "south": "north1",
    "svg": "svg1",
    "uid": "OOOOOOOOOO",
    "west": "east11",
  },
  Object {
    "east": "south1",
    "north": "east11",
    "rotiationDegree": 270,
    "south": "west1",
    "svg": "svg1",
    "uid": "OOOOOOOOOO",
    "west": "north1",
  },
  Object {
    "east": "east2",
    "north": "north2",
    "rotiationDegree": 0,
    "south": "south2",
    "svg": "svg2",
    "uid": "OOOOOOOOOO",
    "west": "west2",
  },
  Object {
    "east": "north2",
    "north": "west2",
    "rotiationDegree": 90,
    "south": "east2",
    "svg": "svg2",
    "uid": "OOOOOOOOOO",
    "west": "south2",
  },
  Object {
    "east": "west2",
    "north": "south2",
    "rotiationDegree": 180,
    "south": "north2",
    "svg": "svg2",
    "uid": "OOOOOOOOOO",
    "west": "east2",
  },
  Object {
    "east": "south2",
    "north": "east2",
    "rotiationDegree": 270,
    "south": "west2",
    "svg": "svg2",
    "uid": "OOOOOOOOOO",
    "west": "north2",
  },
]
`;

exports[`data utils tests toDataFile should combine same tiles 1`] = `
Object {
  "cellSize": 10,
  "tiles": Array [
    Object {
      "east": "east_a",
      "frequency": 1,
      "north": "north_a",
      "rotiationDegrees": Array [
        0,
        90,
        180,
        270,
      ],
      "south": "south_a",
      "svg": "svg_a",
      "west": "west_a",
    },
    Object {
      "east": "east_b",
      "frequency": 2,
      "north": "north_b",
      "rotiationDegrees": Array [
        0,
        90,
      ],
      "south": "south_b",
      "svg": "svg_b",
      "west": "west_b",
    },
    Object {
      "east": "east_c",
      "frequency": 3,
      "north": "north_c",
      "rotiationDegrees": Array [
        0,
      ],
      "south": "south_c",
      "svg": "svg_c",
      "west": "west_c",
    },
  ],
}
`;
